internal/api/hooks.go

```diff
@@ -55,20 +74,176 @@ func (a *API) runPostgresHook(ctx context.Context, tx *storage.Connection, name
 	return response, nil
 }
 
-// invokeHook invokes the hook code. tx can be nil, in which case a new
+func readBodyWithLimit(rsp *http.Response) ([]byte, error) {
+	defer rsp.Body.Close()
+
+	const limit = 20 * 1024 // 20KB
+	limitedReader := io.LimitedReader{R: rsp.Body, N: limit}
+
+	body, err := io.ReadAll(&limitedReader)
+	if err != nil {
+		return nil, err
+	}
+
+	if limitedReader.N <= 0 {
+		// Attempt to read one more byte to check if we're exactly at the limit or over
+		_, err := rsp.Body.Read(make([]byte, 1))
+		if err == nil {
+			// If we could read more, then the payload was too large
+			return nil, fmt.Errorf("payload too large")
+		}
+	}
+
+	return body, nil
+}
+
+func (a *API) runHTTPHook(r *http.Request, hookConfig conf.ExtensibilityPointConfiguration, input, output any) ([]byte, error) {
+	client := http.Client{
+		Timeout: DefaultHTTPHookTimeout,
+	}
+	log := observability.GetLogEntry(r)
+	requestURL := hookConfig.URI
+	hookLog := log.WithFields(logrus.Fields{
+		"component": "auth_hook",
+		"url":       requestURL,
+	})
+
+	inputPayload, err := json.Marshal(input)
+	if err != nil {
+		return nil, err
+	}
+	start := time.Now()
+	for i := 0; i < DefaultHTTPHookRetries; i++ {
+		hookLog.Infof("invocation attempt: %d", i)
+		if time.Since(start) > time.Duration(i+1)*DefaultHTTPHookTimeout {
+			return []byte{}, gatewayTimeoutError(ErrorHookTimeout, "failed to reach hook within timeout")
+		}
+		msgID := uuid.Must(uuid.NewV4())
+		currentTime := time.Now()
+		signatureList, err := crypto.GenerateSignatures(hookConfig.HTTPHookSecrets, msgID, currentTime, inputPayload)
+		if err != nil {
+			return nil, err
+		}
+
+		req, err := http.NewRequest(http.MethodPost, requestURL, bytes.NewBuffer(inputPayload))
```

Comment:
You can enforce the request duration by adding a Context to the request. The context can have a deadline, etc.

It's a very good idea to pass the contexts down to `runHook` so that we get observability for free and that when the toplevel context gets cancelled all the operations that depend on it go away.